src/prototype_creep.js

```diff
@@ -167,9 +167,16 @@ Creep.prototype.stayInRoom = function() {
 };
 
 Creep.prototype.buildRoad = function() {
-  if (this.room.controller && this.room.controller.my && this.pos.lookFor(LOOK_TERRAIN)[0] !== 'swamp' &&
-    (this.room.controller.level < 4 || this.room.memory.misplacedSpawn)) {
-    return false;
+  if (this.room.controller && this.room.controller.my) {
+
+    if (this.pos.lookFor(LOOK_TERRAIN)[0] !== 'swamp' &&
+      (this.room.controller.level < 3 || this.room.memory.misplacedSpawn)) {
+      return false;
+    }
+    // TODO should be extracted to a method, stolen from role_harvester
+    if (!this.room.storage || this.room.storage.store.energy + 1000 < config.creep.energyFromStorageThreshold) {
```

Comment:
I do the same in my workspace too. I mean I also restrict upgrader appetite, but have no clue is we se the same approach)